sql/src/main/java/io/crate/protocols/postgres/BulkPortal.java

```diff
@@ -0,0 +1,227 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.protocols.postgres;
+
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import io.crate.action.sql.ResultReceiver;
+import io.crate.action.sql.RowReceiverToResultReceiver;
+import io.crate.action.sql.SQLOperations;
+import io.crate.analyze.Analysis;
+import io.crate.analyze.Analyzer;
+import io.crate.analyze.ParameterContext;
+import io.crate.concurrent.CompletionListener;
+import io.crate.exceptions.Exceptions;
+import io.crate.executor.Executor;
+import io.crate.executor.TaskResult;
+import io.crate.operation.collect.StatsTables;
+import io.crate.operation.projectors.ResumeHandle;
+import io.crate.planner.Plan;
+import io.crate.planner.Planner;
+import io.crate.sql.tree.Statement;
+import io.crate.types.DataType;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
+import static io.crate.action.sql.SQLRequest.EMPTY_ARGS;
+
+public class BulkPortal implements Portal {
+
+    private String name;
+    private String query;
+    private Statement statement;
+    private SQLOperations.Session session;
+    private Analysis analysis;
+    private int maxRows = 0;
+    private RowReceiverToResultReceiver rowReceiver = null;
+    private List<? extends DataType> outputTypes;
+    private final List<List<Object>> bulkParams = new ArrayList<>();
+    private final List<ResultReceiver> resultReceivers = new ArrayList<>();
+
+    public BulkPortal(String name, SQLOperations.Session session) {
```

Comment:
I'd rather pass in the stuff from the session you need instead of keeping a parent reference.
Will make it easier to create tests for the portal